portfolio/src/main/java/com/google/sps/servlets/ServeCommentsServlet.java

```diff
@@ -45,19 +66,41 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String email = request.getParameter("email");
     String content = request.getParameter("content");
-    long timestamp = System.currentTimeMillis();
+    String userTokenId = extractUserTokenId(request);
+
+    try {
+      GoogleIdToken idToken = validateUserIdToken(userTokenId);
+      System.out.println("Token ID is valid.");
+      addCommentToDb(idToken, content);
+      response.sendRedirect("/leave-comment/leave-comment.html");
+    } catch (GeneralSecurityException e) {
+      System.out.println("Token ID was not valid.");
+      response.sendRedirect("/");
+    }
+  }
 
+  private static String extractUserTokenId(HttpServletRequest request) {
+    String userTokenId = request.getParameter("userTokenId");
+    if (userTokenId == null) {
+      throw new IllegalArgumentException("User token ID is empty.");
+    }
+    return userTokenId;
+  }
+
+  private void addCommentToDb(GoogleIdToken idToken, String content) {
+    String email = idToken.getPayload().getEmail();
+    long timestamp = System.currentTimeMillis();
     Entity commentEntity = new Entity("Comment");
     commentEntity.setProperty("email", email);
     commentEntity.setProperty("content", content);
     commentEntity.setProperty("timestamp", timestamp);
-
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     datastore.put(commentEntity);
+  }
 
-    response.sendRedirect("/leave-comment/leave-comment.html");
+  private GoogleIdToken validateUserIdToken(String idTokenString) throws GeneralSecurityException, IOException {
+    return verifier.verify(idTokenString);
```

Comment:
This file changed a bit, please take a look at new version.
We had a talk with Michał about handling verify() throwing exceptions, because in case of malformed idToken this function is sometimes throwing JSONParsing IOException (prefered result is returning null...) and sometimes returns null. Current solution is some kind of compromise between writing explicitely this exception and ignoring it.